### ISSUE TITLE
[Proposal] Patterns / Additional Case Clauses

### ISSUE THREAD

**Issue by AdamSpeight2008** [dotnet/vblang#24]
[(Ported from Roslyn Repo)](https://github.com/dotnet/roslyn/issues/13440)

Is VB going to support patterns? 
What about guard clauses? eg `Case ... .... When boolEpxr`
What about type matching and value manifesting? eg `TypeIs String As str`
How will tuple clause be handled?

``` vb
Select Case obj
  Case TypeIs String As str When str IsNot Nothing
    ' do stuff
  Case ( TypeIs Boolean As valid, TypeIs Int32 As i32 ) When ( valid AndAlso ( i32 > 0 ) )
   ' manifesting into Tuple(Of Boolean, Int32 ), which has named elements ( valid:= , i32:= ) 
End Select
```

**Comment by ocdtrekkie** [comment]
After reading about this feature in C# 7, I came here looking explicitly for whether or not this functionality would be available in VB. I have a program using an unholy Select Case which would be much cleaner and much happier if I could do this.

As a highlight point, the recent discussion of how VB and C# would diverge going forwards focused on not making VB more complicated to learn, and being able to add a When on a Case is not super complicated (and reads pretty clearly in plain English).
